src/uu/chcon/src/chcon.rs

```diff
@@ -0,0 +1,1130 @@
+                    return libc::EXIT_FAILURE;
+                }
+
+                Ok(Some(true)) | Ok(None) => {}
+            }
+
+            let c_context = if let Ok(value) = os_str_to_c_string(context) {
+                value
+            } else {
+                show_error!("Invalid security context '{}'.", context.to_string_lossy());
+                return libc::EXIT_FAILURE;
+            };
+
+            SELinuxSecurityContext::String(c_context)
+        }
+
+        CommandLineMode::Custom { .. } => SELinuxSecurityContext::default(),
+    };
+
+    let root_dev_ino = if options.preserve_root && options.recursive_mode.is_recursive() {
+        match get_root_dev_ino() {
+            Ok(r) => Some(r),
+
+            Err(r) => {
+                show_error!("{}.", report_full_error(&r));
+                return libc::EXIT_FAILURE;
+            }
+        }
+    } else {
+        None
+    };
+
+    let results = process_files(&options, &context, root_dev_ino);
+    if results.is_empty() {
+        return libc::EXIT_SUCCESS;
+    }
+
+    for result in &results {
+        show_error!("{}.", report_full_error(result));
+    }
+    libc::EXIT_FAILURE
+}
+
+#[derive(thiserror::Error, Debug)]
+enum Error {
+    #[error("No context is specified")]
+    MissingContext,
+
+    #[error("No files are specified")]
+    MissingFiles,
+
+    #[error("{0}")]
+    ArgumentsMismatch(&'static str),
+
+    #[error(transparent)]
+    CommandLine(#[from] clap::Error),
+
+    #[error("{operation} failed")]
+    Io {
+        operation: &'static str,
+        source: io::Error,
+    },
+
+    #[error("{operation} failed on '{}'", .operand1.to_string_lossy())]
+    Io1 {
+        operation: &'static str,
+        operand1: OsString,
+        source: io::Error,
+    },
+}
+
+impl Error {
+    fn io1(operation: &'static str, operand1: impl Into<OsString>, source: io::Error) -> Self {
+        Self::Io1 {
+            operation,
+            operand1: operand1.into(),
+            source,
+        }
+    }
+
+    #[cfg(unix)]
+    fn io1_c_str(operation: &'static str, operand1: &CStr, source: io::Error) -> Self {
+        if operand1.to_bytes().is_empty() {
+            Self::Io { operation, source }
+        } else {
+            use std::os::unix::ffi::OsStrExt;
+
+            Self::io1(operation, OsStr::from_bytes(operand1.to_bytes()), source)
+        }
+    }
+}
+
+#[derive(Debug)]
+struct Options {
+    verbose: bool,
+    dereference: bool,
+    preserve_root: bool,
+    recursive_mode: RecursiveMode,
+    affect_symlink_referent: bool,
+    mode: CommandLineMode,
+    files: Vec<PathBuf>,
+}
+
+static VERSION: &str = env!("CARGO_PKG_VERSION");
+static ABOUT: &str = "Change the SELinux security context of each FILE to CONTEXT. \n\
+                      With --reference, change the security context of each FILE to that of RFILE.";
+
+impl Options {
+    fn command_line_config(usage: &str) -> clap::App {
```

Comment:
I am sorry but could you please follow the structure of other programs?

the clap declaration should be done in uumain

with 60+ programs, consistency is key here
example:
https://github.com/uutils/coreutils/blob/master/src/uu/chown/src/chown.rs#L69-L74

Reply:
I'll change that, but it makes `uumain()` unreadable because it becomes too long.